crates/client-api-messages/src/websocket.rs

```diff
@@ -0,0 +1,326 @@
+    /// The all-zeros address is a sentinel which denotes no address.
+    /// `init` and `update` reducers will have a `caller_address`
+    /// if and only if one was provided to the `publish` HTTP endpoint.
+    /// Scheduled reducers will never have a `caller_address`.
+    /// Reducers invoked by HTTP will have a `caller_address`
+    /// if and only if one was provided to the `call` HTTP endpoint.
+    /// Reducers invoked by WebSocket will always have a `caller_address`.
+    pub caller_address: Address,
+    /// The original CallReducer request that triggered this reducer.
+    pub reducer_call: ReducerCallInfo,
+    /// The amount of energy credits consumed by running the reducer.
+    pub energy_quanta_used: EnergyQuanta,
+    /// How long the reducer took to run.
+    pub host_execution_duration_micros: u64,
+}
+
+/// Contained in a [`TransactionUpdate`], metadata about a reducer invocation.
+#[derive(SpacetimeType, Debug)]
+#[sats(crate = spacetimedb_lib)]
+pub struct ReducerCallInfo {
+    /// The name of the reducer that was called.
+    pub reducer_name: String,
+    /// The numerical id of the reducer that was called.
+    pub reducer_id: u32,
+    /// The arguments to the reducer, encoded as BSATN or JSON according to the reducer's argument schema
+    /// and the client's requested protocol.
+    pub args: EncodedValue,
+    /// An identifier for a client request
+    pub request_id: u32,
+}
+
+/// The status of a [`TransactionUpdate`].
+#[derive(SpacetimeType, Debug)]
+#[sats(crate = spacetimedb_lib)]
+pub enum UpdateStatus {
+    /// The reducer ran successfully and its changes were committed to the database.
+    /// The rows altered in the database/ will be recorded in the `DatabaseUpdate`.
+    Committed(DatabaseUpdate),
+    /// The reducer errored, and any changes it attempted to were rolled back.
+    /// This is the error message.
+    Failed(String),
+    /// The reducer was interrupted due to insufficient energy/funds,
+    /// and any changes it attempted to make were rolled back.
+    OutOfEnergy,
+}
+
+/// A collection of inserted and deleted rows, contained in a [`TransactionUpdate`] or [`SubscriptionUpdate`].
+#[derive(SpacetimeType, Debug, Clone, Default)]
+#[sats(crate = spacetimedb_lib)]
+pub struct DatabaseUpdate {
+    pub tables: Vec<TableUpdate>,
+}
+
+impl DatabaseUpdate {
+    pub fn is_empty(&self) -> bool {
+        self.tables.is_empty()
+    }
+
+    pub fn num_rows(&self) -> usize {
+        self.tables.iter().map(|t| t.num_rows()).sum()
+    }
+}
+
+impl FromIterator<TableUpdate> for DatabaseUpdate {
+    fn from_iter<T: IntoIterator<Item = TableUpdate>>(iter: T) -> Self {
+        DatabaseUpdate {
+            tables: iter.into_iter().collect(),
+        }
+    }
+}
+
+/// Part of a [`DatabaseUpdate`] received by client from database for alterations to a single table.
+#[derive(SpacetimeType, Debug, Clone)]
+#[sats(crate = spacetimedb_lib)]
+pub struct TableUpdate {
```

Comment:
Can this be made generic over the protocol type? If not, I think we should make two types of `TableUpdate`: `TextTableUpdate` and `BinaryTableUpdate`. If we're going to store this information at runtime, best to store it at this level (or higher) rather than *per row*.

Reply:
We can't (currently) use generic types meaningfully in code that goes through `spacetime generate` because computing a `ModuleDef`/`TypeSpace` treats them incorrectly: it emits only the first-encountered specialization, and all later references refer incorrectly to that specialization. So if we had `TableUpdate<Protocol>`, depending on how our code was laid out, we'd get either `struct TableUpdate { rows: RowsJson }` or `struct TableUpdate { rows: RowsBsatn }`, but not both, and all references would refer to the one chosen.

I think we're unlikely to add generics to SATS any time soon, but we could in the near future fix this to emit all used specializations under unique names, like `struct TableUpdate_RowsJson` and `struct TableUpdate_RowsBsatn`, which would eliminate the need for `enum EncodedRow`.

Or, as discussed above, we could just hoist the tag higher up in the message hierarchy.